scipy/optimize/minpack.py

```diff
@@ -541,25 +570,52 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
 
     >>> popt, pcov = curve_fit(func, xdata, ydata)
 
+    Constrain the optimization to the region of ``0 < a < 3``, ``0 < b < 2``
+    and ``0 < c < 1``:
+
+    >>> popt, pcov = curve_fit(func, xdata, ydata, bounds=(0, [3., 2., 1.]))
+
     """
+    if method is None:
+        if bounds is None:
+            method = 'lm'
+        else:
+            method = 'trf'
+
+    if method == 'lm' and bounds is not None:
+        raise ValueError("Method 'lm' only works for unconstrained problems. "
+                         "Use method='trf' or 'dogbox' instead")
+
     if p0 is None:
         # determine number of parameters by inspecting the function
         from scipy._lib._util import getargspec_no_self as _getargspec
         args, varargs, varkw, defaults = _getargspec(f)
         if len(args) < 2:
             msg = "Unable to determine number of fit parameters."
-            raise ValueError(msg)
-        p0 = [1.0] * (len(args)-1)
+            raise ValueError(msg)              
+        if 'self' in args:
+            p0 = np.ones(len(args) - 2)
+        else:
+            p0 = np.ones(len(args) - 1)
```

Comment:
This branching is no longer needed: `_getargspec` removes self from the arg list.